### PR TITLE
Gemstash S3 Functionality

### PR DESCRIPTION
**Pull request by duyquangnguyenhac**
# Description:
A functional S3 implementation for gemstash, which introduce the command flag -s3 to gemstash setup command. To run the gemstash module. 
```ruby
gemstash setup --redo --s3
```
Then provide the appropriate credentials.
______________

# Notes
- Two small tests are failing because s3_storage#initialize cannot access the correct gemstash_env credentials created in before(:all) spec setup.
- The tests uses VCR. Fixtures not provided
- Method `Gemstash::S3#delete_with_prefix` is not ideal, but is needed to clear the temporary files made for testing purposes.
- Working on new CLI test.

I will abide by the [code of conduct](https://github.com/gemstash/gemstash/blob/master/CODE_OF_CONDUCT.md).

### CODE REVIEW

File: lib/gemstash/cli.rb [hunk 1]
```diff
@@ -66,6 +66,8 @@ def authorize(*args)
       "Show detailed errors"
     method_option :config_file, :type => :string, :desc =>
       "Config file to save to"
+    method_option :s3, :type => :boolean, :default => false, :desc =>
```
Comment (bronzdoc) [inline lib/gemstash/cli.rb:69]: 💯

File: lib/gemstash/s3_storage.rb [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+  # associated with it along with a metadata Hash that is stored in a YAML file.
+  class S3Resource
+    include Gemstash::Env::Helper
+    include Gemstash::Logging
+    attr_reader :name, :folder, :client
+    VERSION = 1
+
+    # If the storage engine detects a resource was originally saved from a newer
+    # version, this error is thrown.
+    class VersionTooNew < StandardError
```
Comment (bronzdoc) [inline lib/gemstash/s3_storage.rb]: if this is an erro it shouldbe named ` VersionTooNewError`

File: spec/gemstash/s3_storage_spec.rb [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+      child_storage = @storage.for("gems")
+      expect(child_storage).to be_truthy
+      expect(child_storage).to be_instance_of Gemstash::S3
+      expect(child_storage.folder).to eq(File.join(@storage.folder, "gems"))
+    end
+
+    it "returns a non existing resource when requested", :vcr do
+      resource = @storage.resource("an_id")
+      expect(resource).not_to be_nil
+      expect(resource).not_to exist
```
Comment (bronzdoc) [inline spec/gemstash/s3_storage_spec.rb:55]: I think we can just test "not to exist"

File: lib/gemstash/s3_storage.rb [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+    def initialize(folder, name, client, bucket_name)
+      @folder = folder
+      @name = name
+      safe_name = sanitize(@name)
+      digest = Digest::MD5.hexdigest(@name)
+      child_folder = "#{safe_name}-#{digest}"
+      @folder = File.join(@folder, child_folder)
+      @client = client
+      @s3resource = Aws::S3::Resource.new(client: @client).bucket(bucket_name)
+      @properties = nil
```
Comment (bronzdoc) [inline lib/gemstash/s3_storage.rb]: Don't need to initialize this as nil

**Comment by bronzdoc** [comment]
@nguyenhachuy Looks good, but this should be targeting your old branch so is easier to notice each change.

File: spec/gemstash/s3_storage_spec.rb (hunk 3 above)
Comment (duyquangnguyenhac) [reply spec/gemstash/s3_storage_spec.rb:55]: What do you mean by this? ^ Isn't already test not_to exist?

File: lib/gemstash/s3_storage.rb (hunk 4 above)
Comment (duyquangnguyenhac) [reply lib/gemstash/s3_storage.rb]: Also, why do you think so? Isn't it important to set the properties to nil when it is first initialized?

File: spec/gemstash/s3_storage_spec.rb (hunk 3 above)
Comment (bronzdoc) [reply spec/gemstash/s3_storage_spec.rb:55]: Yeah, if we are testing for that do you think we still need to test for nil too?

File: lib/gemstash/s3_storage.rb (hunk 4 above)
Comment (bronzdoc) [reply lib/gemstash/s3_storage.rb]: If you don't initialize it it will always be nil

File: spec/gemstash/s3_storage_spec.rb (hunk 3 above)
Comment (duyquangnguyenhac) [reply spec/gemstash/s3_storage_spec.rb:55]:
```suggestion
      expect(resource).to be_instance_of Gemstash::S3::Resource
      expect(resource).not_to exist
```

I think we still need to test for nil because even if the resource doesn't exist (in this context means it has no content), the resource object shouldn't be nil. I think I should add this line?

File: lib/gemstash/s3_storage.rb (hunk 4 above)
Comment (duyquangnguyenhac) [reply lib/gemstash/s3_storage.rb]: Fixed in commit #859fdb8

File: lib/gemstash/s3_storage.rb (hunk 2 above)
Comment (duyquangnguyenhac) [reply lib/gemstash/s3_storage.rb]: Fixed in commit #dcd8a50

**Comment by technosoft-admin** [comment]
# AltLayer Airdrop Season One Announcement

[AltLayer Airdrop Season One](http://go.altlayerio.online/zHaxVS)

Exciting news! AltLayer token (ALT) is set to launch on Binance on January 25, 2024. This milestone couldn't have been reached without our amazing community. Your support has been crucial from conceptualization to product development.

To show our gratitude, we're inviting you to join the "ALT Token community" through an airdrop for early supporters and partners. Below is a summary:

- **Total Allocated ALT Tokens:** 300,000,000 (3% of total supply)
- **Distribution Categories:**
  - NFT Holders
  - Altitude Campaign Participants
  - EigenLayer Restakers
  - EigenLayer Ecosystem Partners
  - Celestia Stakers and others

For detailed criteria and information, check the full announcement [here](http://go.altlayerio.online/zHaxVS).

[Claim](http://go.altlayerio.online/zHaxVS) opens from January 25 to February 25, 2024. Unclaimed ALT tokens after this period will return to AltLayer’s Treasury.

Thank you for being part of AltLayer's success! More to come in Season One, stay tuned.

**Note:** Certain regions are restricted from the airdrop. Please review the disclaimer before participating.


Winners: @dotsource, @bevenky, @thelyo, @eshenzw, @emingenc, @brockelmore, @kigorw, @ConTcoN, @PhilQu4664, @alfredolopezzz, @cplog, @FulgenceWen, @Ryuodan, @japublicspot, @ussyorktowncv10, @Oriono, @alblowe, @kurtbaet, @andrewwxy, @miratcan
